### PR TITLE
fix: check all types for toJSON function

### PR DESCRIPTION
**Pull request by ivan-tymoshenko**
@bmenant I've taken your tests from #414 (except two). Can you take a look at this PR?

All meaningful changes are here: https://github.com/fastify/fast-json-stringify/compare/add-tojson-function-handler?expand=1#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R877. All other is necessary refactoring.

Снимок экрана 2022-09-06 в 13 59 15

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

File: test/toJSON.test.js [hunk 1]
```diff
@@ -172,3 +172,1020 @@ test('throw an error on non nullable null object', (t) => {
...
+          type: 'number'
+        }
+      }
+    }
+  }
+
+  const dataNull = { toJSON () { return null } }
+
+  const stringify = build(schema)
+  t.throws(() => stringify({ arr: dataNull }))
```
Comment (Eomm) [inline test/toJSON.test.js:890]: Which error?

File: test/toJSON.test.js [hunk 2]
```diff
@@ -172,3 +172,1020 @@ test('throw an error on non nullable null object', (t) => {
...
+    toJSON () {
+      return { obj: this.props.obj }
+    }
+  }
+
+  const stringify = build(schema)
+  const output = stringify(object)
+
+  JSON.parse(output)
+  t.pass()
```
Comment (Eomm) [inline test/toJSON.test.js:253]: I would move this line into the `toJSON` method, or it seems not informative here

File: test/toJSON.test.js (hunk 1 above)
Comment (ivan-tymoshenko) [reply test/toJSON.test.js:890]: 'The value \'null\' does not match schema definition.'

**Comment by ivan-tymoshenko** [comment]
This feature will cause some performance downgrade. I can add the initial parameter that would turn it off. Should I add it?

**Comment by climba03003** [comment]
> I can add the initial parameter that would turn it off. Should I add it?

Around 5% performance, I think it better turn off by default. Or at least make it configurable.
It can be done on another PR if you like.

**Review by climba03003 — APPROVED** [review]
LGTM.

**Review by Eomm — COMMENTED** [review]
> I think it better turn off by default

Does it not make it braking change setting the option to `false` by default?
Do we want to ship it in the current major?

**Comment by ivan-tymoshenko** [comment]
> > I think it better turn off by default
> 
> 
> 
> Does it not make it braking change setting the option to `false` by default?
> 
> Do we want to ship it in the current major?
> 
> 

Yes, it will be a breaking change if we turn it off by default. That's why I changed the target branch to the next.

I thought that the next major release for the FJS would be possible only with a Fastify major release. Then I don’t see a problem with shipping it. It I’m not right, please explain it to me.

**Review by Eomm — APPROVED** [review]
> Yes, it will be a breaking change if we turn it off by default. That's why I changed the target branch to the next.

Oh, I oversight that!

> I thought that the next major release for the FJS would be possible only with a Fastify major release. Then I don’t see a problem with shipping it. It I’m not right, please explain it to me.

Ni, we can ship a new major for `@fastify/fast-json-schema-serializer`, and then users can use it very easily! So it is not necessary to wait for fastify v5 😄

I'm quite proud of that component, even if it is quite tricky sometimes

**Pull request merged**